### PR TITLE
HANA Provider: Add support for fast extent estimation

### PR DESCRIPTION
**Pull request by jatsqi**
## Description

With QRC/1 2024 a new and faster way to estimate extents of spatial column has been added to HANA.
More precisely, users can now use special columns in [this table](https://help.sap.com/docs/hana-cloud-database/sap-hana-cloud-sap-hana-database-spatial-reference/spatial-extent-with-sys-m-st-geometry-columns?locale=en-US) to get an estimation of the extent of a spatial column.
Even though the stored extent is only an estimation, it is often _good enough_.
Especially for very large tables costly aggregation functions are avoided every time the extent is recomputed.
The extent information in the mentioned HANA table is only available after a delta-merge, so a fallback to the standard behaviour, i.e. via aggregation functions, was added in case the data is unavailable or has not yet been computed.

## Summary of changes:

1. Support for the mentioned fast extent estimation added
2. Fallback in case of unavailability added
3. Checkbox to enable behaviour added to HANA connection widget (see image down below, 'Use estimated table metadata')

image

### CODE REVIEW

File: src/providers/hana/qgshanaprovider.cpp [hunk 1]
```diff
@@ -1336,19 +1337,36 @@ QgsRectangle QgsHanaProvider::estimateExtent() const
...
 
+  if ( QgsHanaUtils::toHANAVersion(conn->getDatabaseVersion()).majorVersion() < 2 )
+    useEstimatedMetadata = false;
+
   try
   {
     QString sql;
-    if ( isSrsRoundEarth( *conn, mSrid ) )
+
+    if ( useEstimatedMetadata )
```
Comment (mrylov) [inline src/providers/hana/qgshanaprovider.cpp:1347]: The try/catch block became even larger. It makes sense to move all code related to the sql string out of this block.

File: src/providers/hana/qgshanaprovider.cpp [hunk 2]
```diff
@@ -1363,12 +1381,24 @@ QgsRectangle QgsHanaProvider::estimateExtent() const
...
+      }
     }
     rsExtent->close();
     return ret;
   }
   catch ( const QgsHanaException &ex )
   {
+    if ( useEstimatedMetadata ) {
+      // Try again without fast extent estimation
+      return estimateExtent( false );
```
Comment (mrylov) [inline src/providers/hana/qgshanaprovider.cpp:1401]: This recursion is dangerous. Why do you expect an exception for the fast extent estimation case?

File: src/providers/hana/qgshanaprovider.cpp [hunk 3]
```diff
@@ -1336,19 +1337,36 @@ QgsRectangle QgsHanaProvider::estimateExtent() const
   if ( conn.isNull() )
     return QgsRectangle();
 
+  if ( QgsHanaUtils::toHANAVersion(conn->getDatabaseVersion()).majorVersion() < 2 )
```
Comment (mrylov) [inline src/providers/hana/qgshanaprovider.cpp]: This is incorrect. We need to check whether it is a HANA Cloud QRC1/2024 or HANA On-Premise SP8.

**Review by mrylov — COMMENTED** [review]
Please add a corresponding test in https://github.com/qgis/QGIS/blob/master/tests/src/python/test_provider_hana.py.
It would also be helpful if you could attach a screenshot of the connection dialog in the PR description.

**Review by strk — COMMENTED** [review]
Shouldn't be there a version check, given the estimated extent is only available starting from a specific version of HANA ?

**Comment by mrylov** [comment]
> Shouldn't be there a version check, given the estimated extent is only available starting from a specific version of HANA ?

Of course, this should be done/adjusted here https://github.com/qgis/QGIS/pull/57654/files/45b199e2c562d4d7bbdc8fc42e1e430053ee4239#diff-83828ab604ff3d901a3d65c5101d059e25501af6bf8588eff656de861423fd55R1340.

File: src/providers/hana/qgshanasettings.cpp [hunk 4]
```diff
@@ -126,6 +127,8 @@ void QgsHanaSettings::setFromDataSourceUri( const QgsDataSourceUri &uri )
     mUserTablesOnly = QVariant( uri.param( QStringLiteral( "userTablesOnly" ) ) ).toBool();
   if ( uri.hasParam( QStringLiteral( "allowGeometrylessTables" ) ) )
     mAllowGeometrylessTables = QVariant( uri.param( QStringLiteral( "allowGeometrylessTables" ) ) ).toBool();
+  if ( uri.hasParam( QStringLiteral( "estimatedmetadata" ) ) )
```
[the remark this answers is not in the document]
Comment (jatsqi) [reply src/providers/hana/qgshanasettings.cpp:130]: If I'm not mistaken, the uri() method of QgsDataSourceUri uses `estimatedmetadata`. I changed it for our local settings, i.e. QgsHanaSettings.

File: src/providers/hana/qgshanasettings.cpp [hunk 5]
```diff
@@ -126,6 +127,8 @@ void QgsHanaSettings::setFromDataSourceUri( const QgsDataSourceUri &uri )
...
+    mUseEstimatedMetadata = QVariant( uri.param( QStringLiteral( "estimatedmetadata" ) ) ).toBool();
```
[the remark this answers is not in the document]
Comment (jatsqi) [reply src/providers/hana/qgshanasettings.cpp:131]: Same as above

File: src/providers/hana/qgshanasettings.cpp [hunk 6]
```diff
@@ -215,6 +219,7 @@ void QgsHanaSettings::load()
     mPassword = settings.value( key + "/password" ).toString();
   mUserTablesOnly = settings.value( key + "/userTablesOnly", true ).toBool();
   mAllowGeometrylessTables = settings.value( key + "/allowGeometrylessTables", false ).toBool();
+  mUseEstimatedMetadata = settings.value( key + "/estimatedmetadata", false ).toBool();
```
[the remark this answers is not in the document]
Comment (jatsqi) [reply src/providers/hana/qgshanasettings.cpp]: Done.

File: src/providers/hana/qgshanaprovider.cpp (hunk 1 above)
Comment (jatsqi) [reply src/providers/hana/qgshanaprovider.cpp]: Done.

File: src/providers/hana/qgshanaprovider.cpp (hunk 2 above)
Comment (jatsqi) [reply src/providers/hana/qgshanaprovider.cpp]: My mistake, I was not sure in which cases the exception is thrown. Removed.

File: src/providers/hana/qgshanaprovider.cpp (hunk 3 above)
Comment (jatsqi) [reply src/providers/hana/qgshanaprovider.cpp]: I added a "proper" version check, which queries the cloud version from SYS.M_DATABASE.

File: src/providers/hana/qgshanaprovider.cpp [hunk 7]
```diff
@@ -1350,7 +1385,10 @@ QgsRectangle QgsHanaProvider::estimateExtent() const
       QString subQuery = buildQuery( QStringLiteral( "ST_EnvelopeAggr(%1) AS ext" ).arg( QgsHanaUtils::quotedIdentifier( mGeometryColumn ) ) );
       sql = QStringLiteral( "SELECT ext.ST_XMin(),ext.ST_YMin(),ext.ST_XMax(),ext.ST_YMax() FROM (%1)" ).arg( subQuery );
     }
+  }
 
+  try
+  {
     QgsHanaResultSetRef rsExtent = conn->executeQuery( sql );
```
Comment (mrylov) [inline src/providers/hana/qgshanaprovider.cpp:1392]: You could restructure the code as follows and get rid of one rsExtent->close() call.

```
    QgsHanaResultSetRef rsExtent = conn->executeQuery( sql );
    QgsRectangle ret;
    if ( rsExtent->next() )
    {
      QVariant val = rsExtent->getValue( 1 );
      if ( !val.isNull() )
      {
        ret.setXMinimum( val.toDouble() );
        ret.setYMinimum( rsExtent->getValue( 2 ).toDouble() );
        ret.setXMaximum( rsExtent->getValue( 3 ).toDouble() );
        ret.setYMaximum( rsExtent->getValue( 4 ).toDouble() );
      }
    }
    rsExtent->close();

   if (useEstimatedMetadata && ret.empty())
        return estimateExtent( false );

```

File: src/ui/qgshananewconnectionbase.ui [hunk 8]
```diff
@@ -890,6 +890,16 @@
         </property>
        </widget>
       </item>
+      <item>
+       <widget class="QCheckBox" name="chkUseEstimatedMetadata">
+        <property name="toolTip">
+         <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;If checked, estimated table metadata will be used if available. For large tables, this avoids a slow table scans and potentially expensive computations, but may result in incorrect layer properties such as layer extent.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (mrylov) [inline src/ui/qgshananewconnectionbase.ui:896]: What if we rephrase this "this avoids a slow table scans" - > "this avoids slow table loads"?

File: src/ui/qgshananewconnectionbase.ui [hunk 9]
```diff
@@ -890,6 +890,16 @@
...
+        </property>
+        <property name="text">
+         <string>Use estimated table metadata</string>
```
[the remark this answers is not in the document]
Comment (jatsqi) [reply src/ui/qgshananewconnectionbase.ui:899]: Done. I included the version for both HANA Cloud and On-Prem.

File: tests/src/python/test_provider_hana.py [hunk 10]
```diff
@@ -539,6 +539,22 @@ def testFilterRectOutsideSrsExtent(self):
         expected = {1, 2, 4, 5}
         self.assertEqual(set(expected), result)
 
+    def testExtentWithEstimatedMetadata(self):
+        create_sql = f'CREATE TABLE "{self.schemaName}"."test_extent" ( ' \
+            'ID INTEGER NOT NULL PRIMARY KEY,' \
+            'GEOM1 ST_GEOMETRY(0),' \
```
[the remark this answers is not in the document]
Comment (jatsqi) [reply tests/src/python/test_provider_hana.py:545]: Done.

File: src/providers/hana/qgshanaprovider.cpp [hunk 11]
```diff
@@ -1336,9 +1357,23 @@ QgsRectangle QgsHanaProvider::estimateExtent() const
   if ( conn.isNull() )
     return QgsRectangle();
 
-  try
+  if ( !checkHANAVersion( *conn, QVersionNumber( 2, 0, 80 ), QVersionNumber( 2024, 2, 0 ) ) )
```
[the remark this answers is not in the document]
Comment (jatsqi) [reply src/providers/hana/qgshanaprovider.cpp]: Good point, done.

File: src/ui/qgshananewconnectionbase.ui (hunk 8 above)
Comment (jatsqi) [reply src/ui/qgshananewconnectionbase.ui]: I like that. Done.

File: src/providers/hana/qgshanaprovider.cpp (hunk 7 above)
Comment (jatsqi) [reply src/providers/hana/qgshanaprovider.cpp:1392]: Done.

File: src/providers/hana/qgshanaprovider.cpp [hunk 12]
```diff
@@ -1327,7 +1328,27 @@ bool QgsHanaProvider::checkPermissionsAndSetCapabilities( QgsHanaConnection &con
   return true;
 }
 
-QgsRectangle QgsHanaProvider::estimateExtent() const
+bool QgsHanaProvider::checkHANAVersion( QgsHanaConnection &conn, const QVersionNumber &premise, const QVersionNumber &cloud ) const
```
[the remark this answers is not in the document]
Comment (jatsqi) [reply src/providers/hana/qgshanaprovider.cpp]: I marked the function as static and removed it from the class.

File: src/ui/qgshananewconnectionbase.ui [hunk 13]
```diff
@@ -890,6 +890,16 @@
         </property>
        </widget>
       </item>
+      <item>
+       <widget class="QCheckBox" name="chkUseEstimatedMetadata">
+        <property name="toolTip">
+         <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;If checked, estimated table metadata will be used if available. For large tables, this avoids slow table loads and potentially expensive computations, but may result in incorrect layer properties such as layer extent. The fast extent estimation is available on HANA Cloud since QRC1/2024 and will be available on HANA On-Premise with SP8 and above.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (mrylov) [inline src/ui/qgshananewconnectionbase.ui:896]: How about to rephrase the last sentence "is available starting with QRC1/2024 and SP8 in HANA Cloud and HANA On-Premise respectively."?

File: src/providers/hana/qgshanautils.cpp [hunk 14]
```diff
@@ -417,6 +417,20 @@ QVersionNumber QgsHanaUtils::toHANAVersion( const QString &dbVersion )
   return QVersionNumber( maj, min, rev );
 }
 
+QVersionNumber QgsHanaUtils::toHANACloudVersion( const QString &dbCloudVersion )
```
Comment (mrylov) [inline src/providers/hana/qgshanautils.cpp:420]: This function is almost identical to the one above. I don't see any value of having two functions. We can combine them in the following way:

```
QVersionNumber QgsHanaUtils::toHANAVersion( const QString &dbVersion )
{
  QString version = dbVersion;
  QStringList strs = version.replace( '-', '.' ).replace( ' ', '.' ).split( '.' );

  if ( strs.length() < 3 )
    return QVersionNumber( 0 );

  const int maj = strs[0].toInt();
  const int min = strs[1].toInt();
  const int rev = strs[2].toInt();
  return QVersionNumber( maj, min, rev );
}
```

**Review by mrylov — COMMENTED** [review]
Thank you @jatsqi.
I have a couple of minor comments.

File: src/providers/hana/qgshanautils.cpp (hunk 14 above)
Comment (jatsqi) [reply src/providers/hana/qgshanautils.cpp]: Done.

File: src/ui/qgshananewconnectionbase.ui (hunk 13 above)
Comment (jatsqi) [reply src/ui/qgshananewconnectionbase.ui]: Sounds better, thanks! Done.

**Comment by mrylov** [comment]
@jatsqi,  the change looks good to me. Many thanks.

Dear QGIS reviewers @rouault, @nyalldawson, @elpaso this PR is ready for review from our side.

**Review by rouault — APPROVED** [review]

**Pull request merged**